File: packages/template-python/src/deepagents_template/runtime/middleware/eviction.py
"""Large output eviction — replaces oversized tool outputs with previews."""

from __future__ import annotations

from typing import Any


def should_evict(content: str, config: Any) -> bool:
    if not getattr(config, "enabled", True):
        return False
    char_per_token = getattr(config, "char_per_token", 4.0)
    token_limit = getattr(config, "token_limit", 20_000)
    estimated_tokens = len(content) / char_per_token
    return estimated_tokens > token_limit


def create_preview(
    content: str, head_lines: int = 5, tail_lines: int = 5
) -> str:
    lines = content.split("\n")
    if len(lines) <= head_lines + tail_lines:
        return content
    head = "\n".join(lines[:head_lines])
    tail = "\n".join(lines[-tail_lines:])
    omitted = len(lines) - head_lines - tail_lines
    return f"{head}\n\n... [{omitted} lines truncated] ...\n\n{tail}"


def create_eviction_middleware(*, config: Any) -> Any:
    """Return an after-tool hook that truncates oversized outputs."""

    def _after_tool(ctx: Any) -> None:
        result = getattr(ctx, "tool_result", None)
        if result is None:
            return
        content = str(result)
        if should_evict(content, config):
            preview = create_preview(
                content,
                head_lines=getattr(config, "head_lines", 5),
                tail_lines=getattr(config, "tail_lines", 5),
            )
            if hasattr(ctx, "tool_result"):
                ctx.tool_result = preview

    return _after_tool

File: packages/template/src/runtime/middleware/eviction.ts
/**
 * Large Output Eviction Middleware
 *
 * Automatically evicts oversized tool outputs to the backend filesystem,
 * replacing them with compact previews. Inspired by pydantic-deepagents'
 * EvictionCapability.
 *
 * Strategy:
 * - Monitor tool output size after each tool call
 * - When output exceeds token threshold, write to backend
 * - Replace original output with preview + file path reference
 * - Agent can read full content via read_file if needed
 *
 * This prevents a single large tool output from blowing up the context window.
 */

import { createMiddleware, ToolMessage } from "langchain";
import type { EvictionConfig } from "../config/config-loader.js";
import { logger } from "../logger.js";

export interface EvictionOptions {
  config: EvictionConfig;
  /** Backend for writing evicted files. Optional — if not provided, eviction is disabled. */
  backend?: {
    write(path: string, content: string): Promise<void>;
  };
}

/**
 * Check if content should be evicted based on size.
 */
export function shouldEvict(
  content: string,
  config: EvictionConfig
): boolean {
  if (!config.enabled) return false;
  const estimatedTokens = content.length / config.charPerToken;
  return estimatedTokens > config.tokenLimit;
}

/**
 * Create a preview of evicted content (head + tail lines).
 */
export function createPreview(
  content: string,
  config: Pick<EvictionConfig, "headLines" | "tailLines">
): string {
  const lines = content.split("\n");

  if (lines.length <= config.headLines + config.tailLines) {
    return content;
  }

  const head = lines.slice(0, config.headLines).join("\n");
  const tail = lines.slice(-config.tailLines).join("\n");
  const omitted = lines.length - config.headLines - config.tailLines;

  return `${head}\n\n... [${omitted} lines truncated] ...\n\n${tail}`;
}

/**
 * Build the replacement message for an evicted tool result.
 */
export function buildEvictedMessage(
  toolCallId: string,
  toolName: string,
  filePath: string,
  preview: string
): ToolMessage {
  return new ToolMessage({
    content: `Tool result too large, saved to: ${filePath}

Read the full result using read_file with offset and limit parameters.
Example: read_file(path="${filePath}", offset=0, limit=100)

Preview (head/tail):

${preview}`,
    tool_call_id: toolCallId,
    name: toolName,
  });
}

/**
 * Create an eviction middleware that handles large tool outputs.
 */
export function createEvictionMiddleware(options: EvictionOptions) {
  const { config, backend } = options;
  const log = logger.child("eviction");

  // Track evicted tool call IDs to prevent double-eviction
  const evictedIds = new Set<string>();

  return createMiddleware({
    name: "eviction",

    beforeAgent: async () => {
      evictedIds.clear();
    },

    wrapToolCall: async (request, handler) => {
      // Execute tool normally
      const result = await handler(request);

      if (!config.enabled || !backend) {
        return result;
      }

      const toolCallId = request.toolCall.id;
      if (!toolCallId || evictedIds.has(toolCallId)) {
        return result;
      }

      // Extract content from result
      let content = "";
      if (result instanceof ToolMessage) {
        content = typeof result.content === "string"
          ? result.content
          : JSON.stringify(result.content);
      } else if (typeof result === "string") {
        content = result;
      } else {
        return result;
      }

      // Check if eviction needed
      if (!shouldEvict(content, config)) {
        return result;
      }

      // Build file path
      const sanitizedId = toolCallId.replace(/[^a-zA-Z0-9_-]/g, "_");
      const filePath = `${config.evictionPath}/${sanitizedId}.txt`;

      try {
        // Write to backend
        await backend.write(filePath, content);
        evictedIds.add(toolCallId);

        // Generate preview
        const preview = createPreview(content, config);

        log.info("Evicted large tool output", {
          tool: request.toolCall.name,
          toolCallId,
          filePath,
          originalChars: content.length,
          previewChars: preview.length,
        });

        // Return replacement message
        return buildEvictedMessage(
          toolCallId,
          request.toolCall.name,
          filePath,
          preview
        );
      } catch (err) {
        log.error("Eviction failed, preserving original content", {
          tool: request.toolCall.name,
          toolCallId,
          error: err instanceof Error ? err.message : String(err),
        });
        // Fallback: return original result
        return result;
      }
    },
  });
}
